src/udev/net/ethtool-util.c

```diff
@@ -206,3 +212,116 @@ int ethtool_set_wol(int *fd, const char *ifname, WakeOnLan wol) {
+        if (r < 0)
+                return -errno;
+
+        if (!info->sset_mask)
+                return -1;
+
+        len = info->data[0];
+
+        strings = malloc0(sizeof(*strings) + len * ETH_GSTRING_LEN);
+        if (!strings)
+                return log_oom();
+
+        strings->cmd = ETHTOOL_GSTRINGS;
+        strings->string_set = stringset_id;
+        strings->len = len;
+
+        ifr->ifr_data = (void *) strings;
+
+        r = ioctl(*fd, SIOCETHTOOL, ifr);
+        if (r < 0)
+                return -errno;
+
+        *gstrings = strings;
+        strings = NULL;
+
+        return 0;
+}
+
+static int find_feature_index(struct ethtool_gstrings *strings, const char *feature) {
+        unsigned i;
+
+        for (i = 0; i < strings->len; i++) {
+                if (streq((char *) &strings->data[i * ETH_GSTRING_LEN], feature))
+                        return i;
+        }
+
+        return -1;
+}
+
+int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
+        _cleanup_free_ struct ethtool_sfeatures *sfeatures = NULL;
+        _cleanup_free_ struct ethtool_gstrings *strings = NULL;
+        int block, bit, i, r;
+        struct ifreq ifr;
+
+        if (*fd < 0) {
+                r = ethtool_connect(fd);
+                if (r < 0)
+                        return log_warning_errno(r, "link_config: could not connect to ethtool: %m");
+        }
+
+        strscpy(ifr.ifr_name, IFNAMSIZ, ifname);
+
+        r = ethtool_get_stringset(fd, &ifr, ETH_SS_FEATURES, &strings);
+        if (r < 0)
+                return log_warning_errno(r, "link_config: could not get ethtool features for %s", ifname);
+
+        sfeatures = malloc0(sizeof(*sfeatures) + DIV_ROUND_UP(strings->len, 32U) * sizeof(sfeatures->features[0]));
+        if (!sfeatures)
+                return log_oom();
+
+        sfeatures->cmd = ETHTOOL_SFEATURES;
+        sfeatures->size = DIV_ROUND_UP(strings->len, 32U);
+
+        for (i = 0; i < _NET_DEV_FEAT_MAX; i++) {
+
+                if (features[i] != -1) {
+
+                        r = find_feature_index(strings, netdev_featute_table[i]);
+                        if (r < 0) {
+                                log_warning_errno(r, "link_config: could not find feature: %s", netdev_featute_table[i]);
+                                continue;
+                        }
+
+                        block = r / 32;
+                        bit = r % 32;
+
+                        if (features[i]) {
+                                sfeatures->features[block].valid |= 1 << bit;
+                                sfeatures->features[block].requested |= 1 << bit;
+                        } else {
+                                sfeatures->features[block].valid  |= 1 << bit;
+                                sfeatures->features[block].requested &= ~(1 << bit);
+                        }
+                }
+        }
+
+        ifr.ifr_data = (void *) sfeatures;
```

Comment:
`src/udev/net/ethtool-util.c: In function ‘ethtool_set_features’:
src/udev/net/ethtool-util.c:312:22: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
         ifr.ifr_data = sfeatures;`